### ISSUE TITLE
Calling AddSqlServer more than once shouldn't result in an error.

### ISSUE THREAD

**Issue by matthewDDennis** [aspnet/EntityFramework#2090]
I am getting the following error when I first use the ContentTaxonomyDbContext.

> An exception of type 'System.InvalidOperationException' occurred in EntityFramework.Core.dll but was not handled in user code
>
> Additional information: The data stores 'SqlServerDataStore' 'SqlServerDataStore' are configured. A context can only be configured to use a single data store.

I have the following method 
```c#
        public static IServiceCollection AddSqlTaxonomyLookups(this IServiceCollection services, string connectionString)
        {
            var entityFrameworkServicesBuilder = new EntityFrameworkServicesBuilder(services);
            entityFrameworkServicesBuilder
                                          .AddSqlServer() // no error if this is commented out
                                          .AddDbContext<ContentTaxonomyDbContext>(options =>
                options.UseSqlServer(connectionString));

            AddLookupServices(services);

            return services;
        }
```
that I call at the end of the Startup.ConfigureServices

```c#
        // This method gets called by the runtime. Use this method to add services to the container.
        public void ConfigureServices(IServiceCollection services)
        {
            // Add Application settings to the services container.
            services.Configure<AppSettings>(Configuration.GetSubKey("AppSettings"));

            // Add EF services to the services container.
            services.AddEntityFramework()
                .AddSqlServer()
                .AddDbContext<CodeProjectIdentityDbContext>(options =>
                    options.UseSqlServer(Configuration["Data:Identity:ConnectionString"]));
             ...
            // Add our services
            services.AddSqlTaxonomyLookups(Configuration["Data:Content:ConnectionString"]);
        }
```

**Comment by matthewDDennis** [comment]
Additionally, I think that calling any *services.AddXXX* extension method multiple times should not be an issue.

In a modular design, a Module might want to call *services.AddEntityFramework(...)* to ensure that it is registered, without knowing if any other code has called it previously.

**Comment by mikary** [comment]
We looked into resolving this exception in #2271, but after some discussion decided it was more appropriate to throw on repeated calls to AddXXX for providers. PR #2330 updates the AddXXX call to throw earlier with a more descriptive error message.

**Comment by matthewDDennis** [comment]
you could just use TryAdd.

-----Original Message-----

**Comment by mikary** [comment]
The AddXXX setup for providers is a bit strange because each provider is expected to provide an implementation type for the ```IDataStoreSource``` service type. When a context is constructed, the appropriate implementation type is selected from the list based on its compatibility with the ```IDbContextOptions```. This isn't quite the scenario TryAdd was designed to accommodate (it only adds to the collection if there isn't an existing registration for the service type).

We were looking into our own implementation pattern / extension for this and leaning toward just throwing (there has been some debate over whether the type of setup you describe is an anti-pattern or a reasonable use case) and for the time being we are opting to throw. There is a potential change to DI (https://github.com/aspnet/DependencyInjection/pull/249) that we are watching, and if it goes through we will probably adopt the pattern instead of throwing. Until we know for sure, this issue will remain open.

**Comment by rynowak** [comment]
@mikary - will have a reaction PR out soon for MVC demonstrating the pattern we were planning to use. Basically what we're hoping for is to make *parts* of the MVC service registration idempotent.

That's not to say that calling `AddMvc()` twice is supported or useful, but it will allow for easier integration with 3rd party components building on top of the lower level parts of the MVC runtime.

**Comment by mikary** [comment]
Fixed in #2670
